kernel/typeops.ml

```diff
@@ -151,27 +151,33 @@ let make_judgev c t =
   Array.map2 make_judge c t
 
 let type_of_apply env func funt argsv argstv =
+  let open CClosure in
   let len = Array.length argsv in
-  let rec apply_rec i typ = 
-    if Int.equal i len then typ
-    else 
-      (match kind (whd_all env typ) with
-      | Prod (_,c1,c2) ->
-	let arg = argsv.(i) and argt = argstv.(i) in
-	  (try
-	     let () = conv_leq false env argt c1 in
-	       apply_rec (i+1) (subst1 arg c2)
-	   with NotConvertible ->
-	     error_cant_apply_bad_type env
-	       (i+1,c1,argt)
-	       (make_judge func funt)
-	       (make_judgev argsv argstv))
-	    
+  let infos = create_clos_infos all env in
+  let tab = create_tab () in
+  let rec apply_rec i typ =
+    if Int.equal i len then term_of_fconstr typ
+    else
+      let typ, _ = whd_stack infos tab typ [] in
+      match fterm_of typ with
+      | FProd (_, c1, c2, e) ->
+        let arg = argsv.(i) in
+        let argt = argstv.(i) in
+        let c1 = term_of_fconstr c1 in
```

Comment:
Two comments:

First, you should include a comment to explain why you know you can drop the stack (it is initially empty).  It turns out to be true that there are no shifts in the stack, I believe; but the reason is not obvious since whd_stack does not call fapp_stack (it is only the case here because [1] e doesn't have any shifts or lifts to start with, and [2] there are never any shifts in the returned stack (or shifts or lifts in substitutions in the returned fconstr) if you only call kni on [1] a stack without shifts and [2] a fconstr whose substitutions have no shifts or lift; which is an invariant that both isn't actually documented anywhere, and wasn't even true before this PR).  Alternatively, you could just uselessly run fapp_stack here.

Second, term_of_fconstr can be rather expensive; it might be nice to delay doing this (especially since conv_leq just turns the constr back into a fconstr) and do a fast path equality on fterms instead.  It's true conversion can be arbitrarily expensive, and the "fast path" equality still has to walk both terms at least once if it succeeds.  But I would think you could still do fast path equality of fterms at lower cost than either full conversion, or converting to to_constr and then doing fast path equality there; and if that failed, the slow path would not be as slow (no need to convert back for conv_leq).  It would be more complicated, though; it would be something similar to the existing conv_leq with all reduction (beta, delta, iota, etc.) disabled (but not identical, because I think that would cause assertions in the current function to throw).